### PR TITLE
[Issue-324] Data loading performance optimization

### PR DESCRIPTION
**Pull request by manishgupta88**
Code Modifications done:

1. Enabled prefetch - code modifications done to make prefetch work according to new code
2. Moved mdkey processing code to a different thread
3. Moved copying of file from local to hdfs as soon as file is completed writing

### CODE REVIEW

File: processing/src/main/java/org/carbondata/processing/mdkeygen/MDKeyGenStep.java [hunk 1]
```diff
@@ -348,12 +353,14 @@ private void initDataHandler() {
     CarbonDataFileAttributes carbonDataFileAttributes =
         new CarbonDataFileAttributes(meta.getTaskNo(), meta.getFactTimeStamp());
     initAggType(msrdataTypes);
+    String carbonDataDirectoryPath = checkAndCreateCarbonStoreLocation();
```
Comment (gvramana) [inline processing/src/main/java/org/carbondata/processing/mdkeygen/MDKeyGenStep.java]: move folder creation to driver

File: processing/src/main/java/org/carbondata/processing/store/CarbonFactDataHandlerColumnar.java [hunk 2]
```diff
@@ -1090,46 +1133,157 @@ private void initialisedataHolder() {
...
+     * @return computed result
+     * @throws Exception if unable to compute a result
+     */
+    @Override public Void call() throws Exception {
+      try {
+        NodeHolder nodeHolder = processDataRows(dataRows);
+        int indexInNodeHolderArray = 0;
+        // if remainder is 0 means data has to be inserted at last index of array
+        // else index will be remainder - 1
+        int remainder = sequenceNumber % numberOfCores;
```
Comment (gvramana) [inline processing/src/main/java/org/carbondata/processing/store/CarbonFactDataHandlerColumnar.java]: use (seq - 1)

File: processing/src/main/java/org/carbondata/processing/store/CarbonFactDataHandlerColumnar.java [hunk 3]
```diff
@@ -1090,46 +1133,157 @@ private void initialisedataHolder() {
...
+    /**
+     * @param nodeHolder
+     * @param index
+     */
+    public synchronized void put(NodeHolder nodeHolder, int index) {
+      nodeHolders[index] = nodeHolder;
+      // notify the consumer thread when index at which object is to be inserted
+      // becomes equal to current index from where data has to be picked for writing
+      if (index == currentIndex) {
+        available = true;
```
Comment (gvramana) [inline processing/src/main/java/org/carbondata/processing/store/CarbonFactDataHandlerColumnar.java]: change it to atomicboolean

**Comment by manishgupta88** [comment]
@gvramana ....handled review comments...please merge

**Comment by gvramana** [comment]
Another Issue #473 is raised, to be handled to track clean up

**Pull request merged**